File: TS/Graphs/src/Traversals/bfs.ts
import Graph from "../types/graph";

function bfs( graph: Graph, source: string ): void {
   
   const queue = [ source ];

   while( queue.length > 0 ) {
      
     const current: string = queue.shift()!;
     console.log(current);

     for( const neighbour of graph[current] ) {
       queue.push( neighbour );
     }
     
   }
}

const graph = {
  a: ["b", "c"],
  b: ["d"],
  c: ["e"],
  d: ["f"],
  e: [],
  f: []
};

bfs( graph, "a");

File: DataStructures/com/graphs/traversals/BFS.java
package com.graphs.traversals;

import com.graphs.model.Graph;

import java.util.*;

public class BFS {

    public static void bfsTraversal(Map<String, ArrayList<String> > graph, String source ) {

        Queue<String> queue = new LinkedList<>();
        queue.offer( source );

        while ( !queue.isEmpty() ) {
             String current = queue.poll();
            System.out.print( current + " ");

            for( String neighbour : graph.get( current )) {
                 queue.offer( neighbour );
            }
        }
    }


    public static void main(String[] args) {

        Graph graph = new Graph();
        bfsTraversal(graph.getGraph(), "a");
    }
}
